pystac/asset.py

```diff
@@ -0,0 +1,165 @@
+from copy import copy
+from typing import Any, Dict, List, Optional, TYPE_CHECKING, Union
+
+import pystac as ps
+from pystac.utils import is_absolute_href, make_absolute_href
+
+if TYPE_CHECKING:
+    from pystac.collection import Collection as Collection_Type
+    from pystac.item import Item as Item_Type
+
+
+class Asset:
+    """An object that contains a link to data associated with an Item or Collection that
+    can be downloaded or streamed.
+
+    Args:
+        href (str): Link to the asset object. Relative and absolute links are both
+            allowed.
+        title (str): Optional displayed title for clients and users.
+        description (str): A description of the Asset providing additional details,
+            such as how it was processed or created. CommonMark 0.29 syntax MAY be used
+            for rich text representation.
+        media_type (str): Optional description of the media type. Registered Media Types
+            are preferred. See :class:`~pystac.MediaType` for common media types.
+        roles ([str]): Optional, Semantic roles (i.e. thumbnail, overview,
+            data, metadata) of the asset.
+        properties (dict): Optional, additional properties for this asset. This is used
+            by extensions as a way to serialize and deserialize properties on asset
+            object JSON.
+
+    Attributes:
+        href (str): Link to the asset object. Relative and absolute links are both
+            allowed.
+        title (str): Optional displayed title for clients and users.
+        description (str): A description of the Asset providing additional details,
+            such as how it was processed or created. CommonMark 0.29 syntax MAY be
+            used for rich text representation.
+        media_type (str): Optional description of the media type. Registered Media Types
+            are preferred. See :class:`~pystac.MediaType` for common media types.
+        properties (dict): Optional, additional properties for this asset. This is used
+            by extensions as a way to serialize and deserialize properties on asset
+            object JSON.
+        owner: The Item or Collection this asset belongs to, or None if it has no owner.
+    """
+
+    def __init__(
+        self,
+        href: str,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+        media_type: Optional[str] = None,
+        roles: Optional[List[str]] = None,
+        properties: Optional[Dict[str, Any]] = None,
+    ) -> None:
+        self.href = href
+        self.title = title
+        self.description = description
+        self.media_type = media_type
+        self.roles = roles
+
+        if properties is not None:
```

Comment:
I prefer explicit None checks as it avoids the edge case where a value evaluates to False but is not None.

Reply:
I think in this case it would be best to use `if not properties`. If `properties` is `False` in this case it's going to get assigned to `self.properties` (since `False is not None`) which we don't want.

The simplest version might be to replace:

```python
if properties is not None:
    self.properties = properties
else:
    self.properties = {}
```

with 

```python
self.properties = properties or {}
```

Reply:
While [this](https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations) says the None case is special, the type for properties is a dict.  So `{}` is `False`, but that leads to just a `{}`.  Anything else is a type violation.  I do find @duckontheweb's suggestion of this quite readable:

```python
self.properties = properties or {}
```

Reply:
Tracked in #322